src/app/util/ember-compatibility-functions.cpp

```diff
@@ -216,218 +216,221 @@ CHIP_ERROR ReadSingleClusterData(FabricIndex aAccessingFabricIndex, const Concre
-                                                         /* write = */ false);
-
-    if (emberStatus == EMBER_ZCL_STATUS_SUCCESS)
-    {
-        EmberAfAttributeType attributeType = metadata->attributeType;
-        bool isNullable                    = metadata->IsNullable();
-        writer                             = attributeDataIBBuilder.GetWriter();
-        VerifyOrReturnError(writer != nullptr, CHIP_NO_ERROR);
-        TLV::Tag tag = TLV::ContextTag(to_underlying(AttributeDataIB::Tag::kData));
-        switch (BaseType(attributeType))
-        {
-        case ZCL_NO_DATA_ATTRIBUTE_TYPE: // No data
-            ReturnErrorOnFailure(writer->PutNull(tag));
-            break;
-        case ZCL_BOOLEAN_ATTRIBUTE_TYPE: // Boolean
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<bool>(*writer, isNullable));
-            break;
-        case ZCL_INT8U_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<uint8_t>(*writer, isNullable));
-            break;
-        case ZCL_INT16U_ATTRIBUTE_TYPE: // Unsigned 16-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<uint16_t>(*writer, isNullable));
-            break;
-        }
-        case ZCL_INT32U_ATTRIBUTE_TYPE: // Unsigned 32-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<uint32_t>(*writer, isNullable));
-            break;
-        }
-        case ZCL_INT64U_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<uint64_t>(*writer, isNullable));
-            break;
-        }
-        case ZCL_INT8S_ATTRIBUTE_TYPE: // Signed 8-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<int8_t>(*writer, isNullable));
-            break;
-        }
-        case ZCL_INT16S_ATTRIBUTE_TYPE: // Signed 16-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<int16_t>(*writer, isNullable));
-            break;
-        }
-        case ZCL_INT32S_ATTRIBUTE_TYPE: // Signed 32-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<int32_t>(*writer, isNullable));
-            break;
-        }
-        case ZCL_INT64S_ATTRIBUTE_TYPE: // Signed 64-bit integer
-        {
-            ReturnErrorOnFailure(attributeBufferToNumericTlvData<int64_t>(*writer, isNullable));
-            break;
+                // TODO: We should probably clone the writer and convert failures here
+                // into status responses, unless our caller already does that.
+                AttributeValueEncoder valueEncoder(writer, aAccessingFabricIndex);
+                ReturnErrorOnFailure(attrOverride->Read(readPath, valueEncoder));
+
+                if (valueEncoder.TriedEncode())
+                {
+                    status = Protocols::InteractionModel::Status::Success;
+                    ExitNow();
```

Comment:
please find a way to not require goto exit. This is a 200-line method (likely too large) and forcing me to jump  context to follow code is hard to read.

Reply:
For supporting list chunking, AttributeValueEncoder will handle encoding of the whole report (since it need to encode multiple list items), then this function can return here with no more actions.

Reply:
I think it is fine to keep ExitNow and file todo here, in the exit, it need to encode dataversion and close container for attributDataIB currently. Song would cleanup this part after listChunk is done.

Reply:
Doesn't it do that? (Encode the dataversion in the code at the exit label?)

Reply:
BTW, it's pretty easy to refactor out the switch statement into its own function, except for one issue: there's no good CHIP_ERROR code to return "out of band" for the unhandled case (in the default clause of the switch), that we can be sure isn't also returned for any of the other error cases in the switch. This can be solved with a separate out parameter to the function. So if it's just the switch that makes this function "too unreadable", that could be handled fairly easily.